### ISSUE TITLE
How to determine list of interesting-offset-ranges for leakage analysis

### ISSUE THREAD

**Issue by taeyoon01** [UzL-ITS/cipherfix#2]
Hi,

I am trying to run the leakage analysis to make sure the hardened binary instrumentation is secure but I don't quite clearly understand how I should determine the list of interesting-offset-ranges to be traced. Is there a suggested way to do this?

Thanks

**Comment by paetscan** [comment]
Hi and thanks for your interest in Cipherfix!

In general, you can include the whole crypto library image in the evaluation. However, there may be false positives in the leakage analysis due to writes to public data that do not get ciphertext side-channel protection. To avoid these possible false positives, you can look for functions in the instrumented binary that are known to be vulnerable (e.g., constant-time swaps) and manually identify the corresponding offsets using tools like `objdump`. It may be helpful to look for the offsets in the original binary, which is a bit more overseeable, and then identify the respective instrumented basic blocks.

In order for the tool to properly align the original and instrumented binaries, the `evalmarker` option must also be set during instrumentation. Please note that the evaluation tool is a bit unstable and error-prone, and may throw misalignment errors due to small inconsistencies in the instrumentation that need to be manually removed from the traces.

**Comment by taeyoon01** [comment]
@paetscan Thanks for the prompt reply.
I ran evaluation.sh on mbedtls (instrumented version: both aes and ecdha / both fast and base-aesrng-evalmarker):
`./evaluate.sh cipherfix_dir/examples/mbedtls/aes cipherfix_dir/examples/mbedtls/aes/instr-fast-aesrng-evalmarker "1.0000.FFFF;4.0000.FFFF" app`

Then, I got the following error for all of the cases I ran on (byte size missing are different but the first line was same): 
`Can not find equivalent instr memtrace line for #90/1060641/90 image 1 offset 3804 (last instr: image 3 offset 951)`
`--> Files don't match`
`    220219 bytes missing from instr memtrace`
`Total 0 blocks`
`Processed 90 lines`

I wonder if this is a misalignment error. Have you experienced the same error? If so, do you have any tips to fixing this error? Thank you.

**Comment by paetscan** [comment]
Yes, this is a misalignment error. For fixing this, you need to open the memtrace files of the original and the instrumented binary. Up to some point, the memtraces are properly aligned. From this point on, you need to check the deviation in a suitable editor (binary trace from taint tracking tool vs. human-readable trace from memwrite tracer). In your example, the error was recognized in image 3 (which should be the dynamic linker), so the actual misalignment happened earlier. These misaligments often arise because of varying control flows in low-level functions, so probably you will need to remove more than one line in the memtrace.
